src/square.cc

```diff
@@ -380,47 +514,124 @@ ModelInstanceState::RequestThread(
+                               "failed to create output buffer in CPU memory"));
+      }
+
+      // Copy IN -> OUT
+      *(reinterpret_cast<int32_t*>(output_buffer)) = element_count;
+
+      // Simulate output delay, if provided.
+      model_state_->get_model_parameters().Sleep(
+          ModelParameters::DelayType::OUTPUT);
     }
 
-    // Copy IN -> OUT
-    *(reinterpret_cast<int32_t*>(output_buffer)) = element_count;
+    // Timestamp at end of the response.
+    uint64_t response_end_ns;
+    SET_TIMESTAMP(response_end_ns);
 
-    // Send the response.
-    LOG_IF_ERROR(
-        TRITONBACKEND_ResponseSend(
-            response, 0 /* flags */, nullptr /* success */),
-        "failed sending response");
+    // Set error for simulated failure.
+    TRITONSERVER_Error* error = nullptr;
+    if (result_type == ModelParameters::InferResultType::FAIL) {
+      error = TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_UNKNOWN, "simulated failure");
+    }
+
+    // Send response if not empty.
+    if (result_type != ModelParameters::InferResultType::EMPTY) {
+      LOG_IF_ERROR(
+          TRITONBACKEND_ResponseSend(response, 0 /* flags */, error),
+          "failed sending response");
+    }
+
+    // Report response statistics.
+    TRITONBACKEND_ModelInstanceResponseStatistics* response_statistics;
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsNew(&response_statistics));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsSetModelInstance(
+            response_statistics, TritonModelInstance()));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsSetResponseFactory(
+            response_statistics, factory.get()));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsSetResponseStart(
+            response_statistics, response_start_ns));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsSetComputeOutputStart(
+            response_statistics, compute_output_start_ns));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(),
+        TRITONBACKEND_ModelInstanceResponseStatisticsSetResponseEnd(
+            response_statistics, response_end_ns));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(), TRITONBACKEND_ModelInstanceResponseStatisticsSetError(
+                           response_statistics, error));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(), TRITONBACKEND_ModelInstanceReportResponseStatistics(
+                           response_statistics));
+    RESPOND_FACTORY_AND_RETURN_IF_ERROR(
+        factory.get(), TRITONBACKEND_ModelInstanceResponseStatisticsDelete(
+                           response_statistics));
```

Comment:
nit: I'd wrap this in a helper function call to keep the flow more concise to read

Reply:
Sure, wrapped into a function.